src/fcmp_pp/fcmp_pp_rust/fcmp++.h

```diff
@@ -0,0 +1,141 @@
+namespace fcmp_pp_rust {
+#include <cstdarg>
+#include <cstdint>
+#include <cstdlib>
+#include <ostream>
+#include <new>
+
+// ----- deps C bindings -----
+
+/// Inner integer type that the [`Limb`] newtype wraps.
+// TODO: test 32-bit platforms
+using Word = uintptr_t;
+
+/// Big integers are represented as an array of smaller CPU word-size integers
+/// called "limbs".
+using Limb = Word;
+
+
+/// Stack-allocated big unsigned integer.
+///
+/// Generic over the given number of `LIMBS`
+///
+/// # Encoding support
+/// This type supports many different types of encodings, either via the
+/// [`Encoding`][`crate::Encoding`] trait or various `const fn` decoding and
+/// encoding functions that can be used with [`Uint`] constants.
+///
+/// Optional crate features for encoding (off-by-default):
+/// - `generic-array`: enables [`ArrayEncoding`][`crate::ArrayEncoding`] trait which can be used to
+///   [`Uint`] as `GenericArray<u8, N>` and a [`ArrayDecoding`][`crate::ArrayDecoding`] trait which
+///   can be used to `GenericArray<u8, N>` as [`Uint`].
+/// - `rlp`: support for [Recursive Length Prefix (RLP)][RLP] encoding.
+///
+/// [RLP]: https://eth.wiki/fundamentals/rlp
+template<uintptr_t LIMBS>
+struct Uint {
+  /// Inner limb array. Stored from least significant to most significant.
+  Limb limbs[LIMBS];
+};
+
+
+/// A residue mod `MOD`, represented using `LIMBS` limbs. The modulus of this residue is constant, so it cannot be set at runtime.
+/// Internally, the value is stored in Montgomery form (multiplied by MOD::R) until it is retrieved.
+template<uintptr_t LIMBS>
+struct Residue {
+  Uint<LIMBS> montgomery_form;
+};
+
+
+/// A constant-time implementation of the Ed25519 field.
+struct SeleneScalar {
+  Residue<32 / sizeof(uintptr_t)> _0;
+};
+static_assert(sizeof(SeleneScalar) == 32, "unexpected size of selene scalar");
+
+
+/// The field novel to Helios/Selene.
+struct HeliosScalar {
+  Residue<32 / sizeof(uintptr_t)> _0;
+};
+static_assert(sizeof(HeliosScalar) == 32, "unexpected size of helios scalar");
+
+struct HeliosPoint {
+  SeleneScalar x;
+  SeleneScalar y;
+  SeleneScalar z;
+};
+
+struct SelenePoint {
+  HeliosScalar x;
+  HeliosScalar y;
+  HeliosScalar z;
+};
+
+// ----- End deps C bindings -----
+
+struct CResult {
+  void* value;
+  void* err;
+};
+
+template<typename T>
+struct Slice {
+  const T *buf;
+  uintptr_t len;
+};
+
+using HeliosScalarSlice = Slice<HeliosScalar>;
+
+using SeleneScalarSlice = Slice<SeleneScalar>;
+
+extern "C" {
+HeliosPoint helios_hash_init_point();
+
+SelenePoint selene_hash_init_point();
+
+uint8_t *helios_scalar_to_bytes(HeliosScalar helios_scalar);
```

Comment:
C linkage with C++ types? I guess this is valid, but I'm not too sure about the rules on this quirk.

Reply:
What type here is being considered C++? The scalars are from Rust with their layout defined in C code above.

Reply:
`HeliosScalar` has an internal type `Residue<32 / sizeof(uint32_t)>` which is a C++ template. This probably works because no fancy "decorators" are needed for the linkage of the outer type, but I'm not certain if the internal C++ type changes anything. At least on the platforms tested so far, it hasn't mattered.

Presumably since C++ has strict layout rules, the `static_assert` on `sizeof` catches the only possible issue - internal padding.

Reply:
Curious if you might have an alternative approach in mind for this code